src/main.rs

```diff
@@ -75,45 +84,58 @@ impl From<AppSW> for Reply {
     }
 }
 
-#[repr(u8)]
-// Instruction set for the app.
-enum Ins {
+/// Possible input commands received through APDUs.
+pub enum Instruction {
     GetVersion,
     GetAppName,
-    GetPubkey,
-    SignTx,
-    Unknown,
+    GetPubkey { display: bool },
+    SignTx { chunk: u8, more: bool },
 }
 
-impl From<ApduHeader> for Ins {
-    fn from(header: ApduHeader) -> Ins {
-        match header.ins {
-            3 => Ins::GetVersion,
-            4 => Ins::GetAppName,
-            5 => Ins::GetPubkey,
-            6 => Ins::SignTx,
-            _ => Ins::Unknown,
+/// APDU parsing logic.
+///
+/// Parses CLA, INS, P1 and P2 bytes to build an [`Ins`]. P1 and P2 are translated to strongly
+/// typed variables depending on the APDU instruction code. Invalid CLA, INS, P1 or P2 values
+/// result in errors with a status word, which are automatically sent to the host by the SDK.
+///
+/// This design allows a clear separation of the APDU parsing logic and commands handling.
+impl TryFrom<ApduHeader> for Instruction {
+    type Error = AppSW;
+
+    fn try_from(value: ApduHeader) -> Result<Self, Self::Error> {
+        match (value.cla, value.ins, value.p1, value.p2) {
+            (CLA, 3, 0, 0) => Ok(Instruction::GetVersion),
+            (CLA, 4, 0, 0) => Ok(Instruction::GetAppName),
+            (CLA, 5, 0 | 1, 0) => Ok(Instruction::GetPubkey {
+                display: value.p1 != 0,
+            }),
+            (CLA, 6, P1_SIGN_TX_START, P2_SIGN_TX_MORE)
+            | (CLA, 6, 1..=P1_SIGN_TX_MAX, P2_SIGN_TX_LAST | P2_SIGN_TX_MORE) => {
+                Ok(Instruction::SignTx {
+                    chunk: value.p1,
+                    more: value.p2 == P2_SIGN_TX_MORE,
+                })
+            }
+            (CLA, 3..=6, _, _) => Err(AppSW::WrongP1P2),
+            (CLA, _, _, _) => Err(AppSW::InsNotSupported),
+            (_, _, _, _) => Err(AppSW::ClaNotSupported),
         }
     }
 }
 
 // Developer mode / pending review popup
 // must be cleared with user interaction
 fn display_pending_review(comm: &mut Comm) {
-    use ledger_device_sdk::buttons::ButtonEvent::{
```

Comment:
I put these imports here so that the developers only have to remove this function to clean the code when the review is finished (and not have to worry about imports at the top of the file). 

Can't we let it that way ?

Reply:
Ok my bad I did not understood it that way and found the code more readable with the imports at the top. I will revert this change.
For the future maybe we could have a function in the SDK that does this.